File: supabase/functions/export-timesheet/timesheet_employee_visibility.ts
/**
 * Правила видимости строк табеля (сетка UI и Excel).
 *
 * Держите в синхронизации с
 * `lib/features/timesheet/domain/timesheet_employee_visibility.dart`.
 *
 * Сегмент «С часами / Без часов» в Excel не применяется — только базовые правила
 * (объекты, уволенные с часами) и опционально `onlyEmployeeIds` из чекбоксов.
 */

export interface TimesheetHoursIndex {
  employeeIdsWithEntries: Set<string>;
  hoursSumByEmployeeId: Map<string, number>;
}

export function buildTimesheetHoursIndex(
  entries: { employeeId: string; hours: number }[],
): TimesheetHoursIndex {
  const employeeIdsWithEntries = new Set<string>();
  const hoursSumByEmployeeId = new Map<string, number>();

  for (const entry of entries) {
    employeeIdsWithEntries.add(entry.employeeId);
    hoursSumByEmployeeId.set(
      entry.employeeId,
      (hoursSumByEmployeeId.get(entry.employeeId) ?? 0) + entry.hours,
    );
  }

  return { employeeIdsWithEntries, hoursSumByEmployeeId };
}

export function isTimesheetGridEmployeeVisible(
  employee: { id: string; status: string; includeInTimesheet: boolean },
  hoursIndex: TimesheetHoursIndex,
  hasObjectFilter: boolean,
): boolean {
  if (hasObjectFilter) {
    return hoursIndex.employeeIdsWithEntries.has(employee.id);
  }
  if (!employee.includeInTimesheet) {
    return hoursIndex.employeeIdsWithEntries.has(employee.id);
  }
  if (employee.status !== "fired") return true;
  return hoursIndex.employeeIdsWithEntries.has(employee.id);
}

export function filterTimesheetGridEmployees<T extends {
  id: string;
  status: string;
  includeInTimesheet: boolean;
  fullName: string;
}>(
  employees: T[],
  hoursIndex: TimesheetHoursIndex,
  hasObjectFilter: boolean,
  onlyEmployeeIds?: string[],
): T[] {
  let visible = employees
    .filter((employee) =>
      isTimesheetGridEmployeeVisible(employee, hoursIndex, hasObjectFilter)
    )
    .sort((a, b) => a.fullName.localeCompare(b.fullName, "ru"));

  if (onlyEmployeeIds && onlyEmployeeIds.length > 0) {
    const pick = new Set(onlyEmployeeIds);
    visible = visible.filter((e) => pick.has(e.id));
  }

  return visible;
}

File: lib/features/timesheet/domain/timesheet_employee_visibility.dart
import 'package:projectgt/core/utils/formatters.dart';
import 'package:projectgt/domain/entities/employee.dart';

import 'timesheet_employee_list_scope.dart';
import 'timesheet_hours_index.dart';

/// Правила видимости строк табеля (сетка UI и Excel).
///
/// **Единый контракт** для экрана и `export-timesheet`. Держите в синхронизации
/// с `supabase/functions/export-timesheet/timesheet_employee_visibility.ts`.
///
/// Сегмент [TimesheetEmployeeListScope] (`С часами` / `Без часов`) — **только UI**;
/// экспорт Excel его не применяет (всегда [TimesheetEmployeeListScope.all]).

/// Показывать ли сотрудника в сетке/Excel по базовым правилам.
///
/// Объекты, мягкое исключение из табеля (`includeInTimesheet == false`) и уволенные:
/// без записей часов за период — строка скрыта (кроме режима «только с объектами»).
bool isTimesheetGridEmployeeVisible({
  required bool isFired,
  required bool includeInTimesheet,
  required String employeeId,
  required TimesheetHoursIndex hoursIndex,
  required bool hasObjectFilter,
}) {
  if (hasObjectFilter) {
    return hoursIndex.employeeIdsWithEntries.contains(employeeId);
  }
  if (!includeInTimesheet) {
    return hoursIndex.employeeIdsWithEntries.contains(employeeId);
  }
  if (!isFired) return true;
  return hoursIndex.employeeIdsWithEntries.contains(employeeId);
}

/// Применяет сегмент «Все / С часами / Без часов» (только UI).
List<Employee> filterEmployeesByTimesheetListScope(
  List<Employee> employees,
  TimesheetHoursIndex hoursIndex,
  TimesheetEmployeeListScope listScope,
) {
  switch (listScope) {
    case TimesheetEmployeeListScope.withHours:
      final withHours = hoursIndex.employeeIdsWithPositiveHours;
      return employees.where((e) => withHours.contains(e.id)).toList();
    case TimesheetEmployeeListScope.withoutHours:
      return employees
          .where((e) => hoursIndex.hoursSumFor(e.id) <= 0)
          .toList();
    case TimesheetEmployeeListScope.all:
      return employees;
  }
}

List<Employee> _applyListScope(
  List<Employee> employees,
  TimesheetHoursIndex hoursIndex,
  TimesheetEmployeeListScope listScope,
) =>
    filterEmployeesByTimesheetListScope(employees, hoursIndex, listScope);

/// Список сотрудников для строк сетки табеля (без поиска по ФИО).
List<Employee> visibleTimesheetGridEmployees({
  required List<Employee> employees,
  required TimesheetHoursIndex hoursIndex,
  required bool hasObjectFilter,
  TimesheetEmployeeListScope listScope = TimesheetEmployeeListScope.all,
}) {
  final base = employees
      .where(
        (e) => isTimesheetGridEmployeeVisible(
          isFired: e.status == EmployeeStatus.fired,
          includeInTimesheet: e.includeInTimesheet,
          employeeId: e.id,
          hoursIndex: hoursIndex,
          hasObjectFilter: hasObjectFilter,
        ),
      )
      .toList();

  final scoped = _applyListScope(base, hoursIndex, listScope);

  scoped.sort((a, b) {
    final nameA = formatFullName(a.lastName, a.firstName, a.middleName);
    final nameB = formatFullName(b.lastName, b.firstName, b.middleName);
    return nameA.compareTo(nameB);
  });

  return scoped;
}

/// Сотрудники для Excel: базовые правила + опционально только выбранные id.
///
/// [onlyEmployeeIds] — чекбоксы в сетке; `null` или пусто — без ограничения.
List<Employee> visibleTimesheetExportEmployees({
  required List<Employee> employees,
  required TimesheetHoursIndex hoursIndex,
  required bool hasObjectFilter,
  List<String>? onlyEmployeeIds,
}) {
  var visible = visibleTimesheetGridEmployees(
    employees: employees,
    hoursIndex: hoursIndex,
    hasObjectFilter: hasObjectFilter,
    listScope: TimesheetEmployeeListScope.all,
  );

  if (onlyEmployeeIds != null && onlyEmployeeIds.isNotEmpty) {
    final pick = onlyEmployeeIds.toSet();
    visible = visible.where((e) => pick.contains(e.id)).toList();
  }

  return visible;
}
